// intercom-client/dist/bulk.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var Bulk = (function () {
  function Bulk(client, dataType) {
    _classCallCheck(this, Bulk);

    this.client = client;
    this.dataType = dataType;
  }

  _createClass(Bulk, [{
    key: 'bulk',
    value: function bulk(bulkOperations, f) {
      var dataType = this.dataType;
      var bulkParams = { items: [] };
      var url = '/bulk/' + dataType + 's';
      bulkOperations.forEach(function (params) {
        for (var k in params) {
          var data = params[k];
          var method = k === 'create' ? 'post' : k;
          if (method !== 'post' && method !== 'delete') {
            throw new Error('Unknown bulk type: ' + method);
          }
          bulkParams.items.push({
            method: method,
            data_type: dataType,
            data: data
          });
        }
      });
      return this.client.post(url, bulkParams, f);
    }
  }]);

  return Bulk;
})();

exports['default'] = Bulk;
module.exports = exports['default'];
